src/EventStore.Core/DataStructures/ProbabilisticFilter/MemoryMappedFileBloomFilter/MemoryMappedFileBloomFilter.cs

```diff
@@ -0,0 +1,275 @@
+		public const long MinSizeKB = 10;
+		public const long MaxSizeKB = 4_000_000;
+		public const double RecommendedFalsePositiveProbability = 0.02;
+		private const int CacheLineSize = BloomFilterIntegrity.CacheLineSize;
+		private const int NumHashFunctions = 6;
+		private readonly long _numBits;
+		private readonly Header _header;
+		private readonly byte* _pointer = null;
+
+		private readonly IHasher[] _hashers = {
+			new XXHashUnsafe(seed: 0xC0015EEDU),
+			new Murmur3AUnsafe(seed: 0xC0015EEDU),
+			new XXHashUnsafe(seed: 0x13375EEDU),
+			new Murmur3AUnsafe(seed: 0x13375EEDU)
+		};
+
+		private readonly FileStream _fileStream;
+		private readonly long _fileSize;
+		private readonly MemoryMappedFile _mmf;
+		private readonly MemoryMappedViewAccessor _mmfWriteAccessor;
+		private readonly object _writeLock = new();
+
+		private bool _disposed;
+
+		public int CorruptionRebuildCount => _header.CorruptionRebuildCount;
+
+		/// <summary>
+		/// Bloom filter implementation which uses a memory-mapped file for persistence.
+		/// </summary>
+		/// <param name="path">Path to the bloom filter file</param>
+		/// <param name="size">Size of the bloom filter in bytes</param>
+		///
+		/// Add and MightContain can be called concurrently on different threads.
+		/// If the call to Add is complete then a call to MightContain will find the added value.
+		public MemoryMappedFileBloomFilter(string path, bool create, long size, int corruptionRebuildCount = 0) {
+			Ensure.NotNull(path, nameof(path));
+
+			if (size < MinSizeKB * 1000 || size > MaxSizeKB * 1000) {
+				throw new ArgumentOutOfRangeException(nameof(size), $"size should be between {MinSizeKB:N0} and {MaxSizeKB:N0} KB inclusive");
+			}
+
+			_numBits = size * 8;
+
+			_fileStream = new FileStream(
+				path,
+				// todo: OpenOrCreate can be CreateNew if the mininode closed the file before declaring itself 'stopped'
+				create ? FileMode.OpenOrCreate : FileMode.Open,
+				FileAccess.ReadWrite,
+				FileShare.ReadWrite);
+
+			if (Header.Size > CacheLineSize)
+				throw new Exception("header needs to fit in a cache line");
+
+			// add room for filter bits, rounded up to nearest cacheline
+			_fileSize = GetBytePositionInFile(_numBits) / CacheLineSize * CacheLineSize;
+			_fileSize += CacheLineSize;
+			_fileStream.SetLength(_fileSize);
+
+			_mmf = MemoryMappedFile.CreateFromFile(
+				fileStream: _fileStream,
+				mapName: null,
+				capacity: 0,
+				access: MemoryMappedFileAccess.ReadWrite,
+				inheritability: HandleInheritability.None,
+				leaveOpen: false);
+
+			_mmfWriteAccessor = _mmf.CreateViewAccessor(CacheLineSize, 0, MemoryMappedFileAccess.ReadWrite);
+			_mmfWriteAccessor.SafeMemoryMappedViewHandle.AcquirePointer(ref _pointer);
+
+			if (create) {
+				FillWithZeros();
+
+				_header = new Header() {
+					Version = Header.CurrentVersion,
+					CorruptionRebuildCount = corruptionRebuildCount,
+					NumBits = _numBits
+				};
+
+				_header.WriteTo(_mmf);
```

Comment:
Do we need to force a flush here to make sure the zeros and the header are definitely on disk before starting to use the file? It would be somewhat unlucky to have a hardware crash in between, but possible I guess

Reply:
the checkpoint is only flushed after a successful flush of the filter, so if we have a checkpoint on start we are guaranteed that the header and zeros wrote correctly. if we don't have a checkpoint then the filter is junked and rebuilt